csrc/device_lower/utils.cpp

```diff
@@ -1948,24 +1946,38 @@ Val* proveLinearAndGetStride(
     return linear_g->front()->fusion()->zeroVal();
   }
   // Propagate from linear_g to domain. Use frontier to keep track of the
-  // how linear_g lives in the current propagation front.
+  // how linear_g lives in the current propagation front. Note that linear_g may
+  // not contain full dependency of domain.
   Projection frontier = linear_g;
   auto path =
-      ValGraphBFS::getExprGroupsBetween(id_graph, domain, {linear_g}).first;
-  while (!path.empty()) {
-    const auto& [eg, direction] = path.back();
-    path.pop_back();
-    auto from = fromGroups(id_graph, eg, direction);
+      ValGraphPermissiveBFS::getExprGroupsBetween(
+          id_graph, {linear_g}, domain, /*require_all_to_visited=*/false)
+          .first;
+  // Try to prove linear and get stride with the current frontier. We may not
+  // have finished the propagation yet, but because ValGraphPermissiveBFS
+  // supports missing dependency, and with missing dependency, we only have
+  // partial information on how to reach to a state that is easiest for our
+  // proof. It is possible that the easiest state is not the final state of
+  // the propagation. So we need to try the proof each step of the
+  // propagation.
```

Comment:
Could you give a few examples here? Specifically, one for a case where no propagation is done, another that involves a partial traversal of `path`, and also a case where complete traversal of `path` is required.

Reply:
I completely rewrite the comment here, and added a new test `ProveLinearAndGetStrideEarlyStopping` demonstrating why early stopping is important. I believe the new comment and the single example included in it is sufficiently to clearly explain the reason for early stopping. Let me know if you believe more examples are needed.